src/compiler/interface.jl

```diff
@@ -171,6 +187,40 @@ function copy!(x::AbstractVector,  gs::Grads)
   x
 end
 
+broadcasted(f, gs::Grads, gss::ADictOrGrads...) = map(f, gs, gss...)
+
+broadcasted(f, a::Numeric, gs::Grads) = map(x -> f(a, x), gs)
+broadcasted(f, gs::Grads, a::Numeric) = map(x -> f(x, a), gs)
+
+function materialize!(gs1::Grads, gs2::Grads)
+  issetequal(gs1.params, gs2.params) || 
+    throw(ArgumentError("Expected Grads objects with the same Params."))
+  for p in gs1.params
+    gs1[p] = gs2[p]
+  end
+  return gs1
+end
+
+
+function Base.map(f, gs1::Grads, gss::ADictOrGrads...)
+  gsout = Grads(IdDict{Any,Any}(), Params(gs1.params))
+  return map!(f, gsout, gs1, gss...)
+end
+
+function Base.map!(f, gsout::Grads, gss::ADictOrGrads...)
+  all(issetequal(gsout.params, keys(gs)) for gs in gss) || 
+    throw(ArgumentError("map! expects Grads objects with the same Params."))
+  for p in gsout.params
+    gsout[p] = f((_getformap(gs, p) for gs in gss)...) 
+  end
+  return gsout
+end
+
+function _getformap(gs, p)
+  g = gs[p]
+  isnothing(g) ? fill!(similar(p), 0) : g 
```

Comment:
```suggestion
  isnothing(g) ? fill!(similar(p), zero(g)) : g 
```
unless the promoted type is inferred?

Reply:
`fill!` doesn't promote

Reply:
Shouldn't the type returned by `fill!` match the type of `g` for stability?

Reply:
It should I think. Could it be a footgun performance wise?

Reply:
I think you guys are doing multiple mistakes here. `fill!` return type depends only on the first argument, so `typeof(p)`. `fill!` is called only when `g == nothing`, so deriving its return type from `g` is meaningless. The branching based on a nothing check is normally optimized away. The function is probably type-unstable, but the cause is upstream, in Grads containing an `IdDict{Any}`.